staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -309,6 +334,39 @@ func (w *watchCache) updateCache(event *watchCacheEvent) {
 	w.endIndex++
 }
 
+// Use dynamic size cache. If all cached events occur inside eventFreshDuration, increase cache capacity by 2x.
+// Decrease cache capacity by 2x when recent half events occur outside eventFreshDuration.
+func (w *watchCache) resizeCache(eventTime time.Time) {
+	if w.endIndex == w.startIndex+w.capacity && eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
+		capacity := min(w.capacity*2, w.upperBoundCapacity)
+		if capacity > w.capacity {
+			w.makeNewCache(capacity)
+			value.RecordWatchCacheCapacityIncrease()
+			w.capacity = capacity
+		}
+		return
+	}
+	if w.endIndex-w.startIndex > w.capacity/2 && eventTime.Sub(w.cache[(w.endIndex-w.capacity/2)%w.capacity].RecordTime) > eventFreshDuration {
```

Comment:
I would actually change this condition a bit. If we increased the capacity to w.capacity at some point, before decreasing it I would actually wait until it's full. So I would actually change the first part of the condition to:
```
w.endIndex == w.startIndex+w.capacity
```

Having said that, maybe introduce a helper method:
```
func (w *watchCache) isCacheFullLocked() bool {
  return w.endIndex == w.startIndex+w.capacity
}
```
and use it both here and above?

Reply:
Done.